### DISCUSSION TITLE
Synergy with longform

### DISCUSSION THREAD

**Discussion by arussbach** [chuangcaleb/obsidian-fountain-editor#26] [Tips & Tricks]
Hi, I've been using this plugin for two days in combination with the longform plugin [https://github.com/kevboh/longform](url). I write every scene as a note and compile them in a single note with longform. Longform lets you custumise compilation steps, I don't know javascrip but I managed to create a very simple step to add the foutnain header to the final compiled note, so I can use the compiled note with the fountain plugin right away. It's not much, but I thought I would share.

1. Save this code in a .js file stored in your vault.

```
async function compile(input, context) {
      return {
        ...input,
        contents: 
        "---\ncssclasses:\n  - fountain\n---\n" + input.contents
      };
  }
  
  module.exports = {
    description: {
      name: "Fountain Header",
      description: "Add a fountain header to the manuscript",
      availableKinds: ["Manuscript"],
      options: []
    },
    compile: compile
  };
```

2. Tell the longform plugin to look for user scripts in the folder you saved your file in.
3. In the complie tab of longform, add an extra step at the end with your custum step.

If some of you know Js and make steps to number scene headers, or to change the output file format to .fountain. Please share them, it would be great.

**Comment by chuangcaleb** [comment]
![it ain't much, but it's honest work.](https://media.tenor.com/sH_KNNF07EoAAAAC/honest-word-its-honest-work.gif)

Thanks so much for this community contribution! 

Btw, in the `Save as Note` step, you can save with a `.fountain` extension! I explicitly requested for this!
[Feature Request: Save as Note with file extensions other than `.md` · Issue #203 · kevboh/longform](https://github.com/kevboh/longform/issues/203)

**Reply by arussbach** [reply]
Ah, great, thx for doing the request !
But has it already been implementend ?
If I enter the name of my project followed by .fountain, I get someting like "filename.fountain.md". Am I doing something wrong?

**Reply by chuangcaleb** [reply]
Hmm it is working for me. I'm on the latest Longform version v2.0.4, what version of Longform are you on?

**Reply by arussbach** [reply]
That is strange.
I'm on 2.0.4 too.
Here is what I fill in the path : "../Exports/$1.fountain" and I get "title.fountain.md".

**Comment by arussbach** [comment]
Anding on the topic of working with longform, how do you guys think the best way to organise scenes and notes is?

I've just started prepending each scene with a commented area that with all the "meta-information" for the scene. Looks more or less like this:

/*
###### Summary
Stuff happends
###### Characters
- [[Max]]
- [[Nora]]
###### Setting
[[Hotel]] [[Restaurant]]
###### Notes
- 
###### TODO
- [ ] 
*/

INT. HOTEL - DAY

And here the scene.

I'm hesitating on moving these extra informations on a external note and just have a link to it. It would look like :

= ![[Hotel scene infos]]

INT. HOTEL - DAY

And here the scene.

It would be more "atomic" but I'm scared to end up with an overwhelming number of notes to manage. I'm super new to Obsidian, but from what I gathered, having tons of notes is not such a bad thing, am I undersanding that correctly ?

And how do you organise your projects with longform ?

**Reply by chuangcaleb** [reply]
I've personally tried moving away as much as possible 
- from having meta information in the document body, 
- to putting it in the Properties/frontmatter metadata section (now that it's more maturely developed)

This forces me to think of metadata as key-value pairs, which
- is better organization structure 
- is also more effective for search/dataview queries.

So "Characters" and Settings" could go into that frontmatter section as the List type, a list of links.

However, long form prose stuff like Notes and TODO lists probably should still be in the document body.

For one-liner metadata like Summary, I've been using inline dataview fields (`field:: value`).

---

Regarding "extra infos" and note management, I think that's really up to each project and person.

If you want to have a "Meta" note for each scene, you could adopt a parent-child system, where each Fountain-md note has a `meta` property that links to the meta note, while each meta note has a `product` property that links to the Fountain-md note.

The way I'd do it is to try to keep it in one note first, then it's always easier to split out later if the meta/notes section gets too long — this method signals to you that if there's no meta note for a particular scene, then there's probably not too much extra attention needed for that scene. So some scenes have additional meta notes, some don't.

---

Organizing with longform is really up to you, also depends on which stage of writing development you are on (how much granularity you need), and how large the total product would come up to. And when you reorder scenes or sequences or whatever: wherever you define those chunks as swappable, there is where you define your "unit". Though if I'm doing a high-level beat sheet draft, I'd probably keep it quite high-level.

Again I think a rule of thumb I've found useful is that breaking/splitting/fracturing is easier/simpler/more intuitive than assembling/combining. But it'll take a while for each person to find their groove with Obsidian!

**Reply by arussbach** [reply]
Thanks for sharing ! A lot of things I didn't know about yet. Looking forward to explore frontmatter and properties more.
I started having multiple files per scene and I must say it was a bit hellish to maintain (my project is 170 scenes 😨). I quickly reverted to having all data in the scenes themself. I think your idea of splitting things as later as possible is a good one.

**Comment by arussbach** [comment]
Hi,
I have created another script for the longform compiler. It is made to automaticaly number scenes. You have the option to preserve existing scene numbers or override them.

Warning : This script doesn't account for other scen numbering formats than the simpe #1# type, if you use #1A# or other foutain scene numberings, their won't work. If someone wants to accout for these cases you're welcome to modifie the script.
```

// Detect and number Fountain headers
async function compile(input, context) {
  // Regular expression to match Fountain headers
  const headerRegex =
    /^((?:\*{0,3}_?)?(?:(?:int|ext|est|i\/e|int\/ext)[. ]).+)|^(?:\.(?!\.+))(.+)/i;

  // Counter to keep track of header numbers
  let headerNumber = 1;

  // Collection to store existing scene numbers
  const existingSceneNumbers = new Set();

  // Get user option to preserve existing numbering
  const preserveExistingNumbering =
    context.optionValues['preserve-existing-numbering'] || false;

  // Extract existing scene numbers as integers from the manuscript contents if preserving existing numbering
  if (preserveExistingNumbering) {
    const existingNumberMatches = input.contents.match(/#\d+#/g);
    if (existingNumberMatches) {
      existingNumberMatches.forEach(match => {
        const existingNumber = parseInt(match.match(/\d+/)[0]);
        existingSceneNumbers.add(existingNumber);
      });
    }
  }

  // Process each line in the manuscript contents
  const updatedLines = input.contents.split('\n').map(line => {
    // Check if the line matches the Fountain header pattern
    const match = line.match(headerRegex);
    if (match) {
      // If preserving existing numbering and the header already has a number, skip numbering
      if (preserveExistingNumbering && /\s*#\d+#$/.test(line)) {
        return line;
      }

      // Remove existing header number if not preserving
      const lineWithoutNumber = preserveExistingNumbering
        ? match[0]
        : match[0].replace(/#\d+#$/, '');

      // Ensure the headerNumber is not in existingSceneNumbers
      while (existingSceneNumbers.has(headerNumber)) {
        headerNumber++;
      }

      // Create the numbered header
      const numberedHeader = `${lineWithoutNumber}  #${headerNumber}#`;

      // Add the new headerNumber to existingSceneNumbers
      existingSceneNumbers.add(headerNumber);

      // Increment the header number for the next header
      headerNumber++;

      // Return the line with the numbered header
      return numberedHeader;
    }

    // Return the line unchanged if it's not a header
    return line;
  });

  // Join the updated lines to form the modified manuscript contents
  const updatedContents = updatedLines.join('\n');

  // Return the modified manuscript
  return {
    ...input,
    contents: updatedContents,
  };
}

// Export the script configuration
module.exports = {
  description: {
    name: 'Number Fountain scene Headers',
    description: 'Numbers Fountain scene headers',
    availableKinds: ['Manuscript'],
    options: [
      {
        id: 'preserve-existing-numbering',
        name: 'Preserve Existing Numbering',
        description: 'Preserve existing scene numbering if present.',
        type: 'Boolean',
        default: false,
      },
    ],
  },
  compile: compile,
};

```
